src/mpid/ch4/netmod/ofi/ofi_init.c

```diff
@@ -1249,107 +1098,414 @@ int MPIDI_OFI_create_intercomm_from_lpids(MPIR_Comm * newcomm_ptr, int size, con
+    struct fid_cq *cq;
+
+    struct fid_ep *ep;
+    struct fid_ep *tx;
+    struct fid_ep *rx;
+
+#ifdef MPIDI_OFI_VNI_USE_DOMAIN
+    mpi_errno = create_vni_domain(&domain, &av, &rma_cmpl_cntr);
+    MPIR_ERR_CHECK(mpi_errno);
+    mpi_errno = create_cq(domain, &cq);
+    MPIR_ERR_CHECK(mpi_errno);
+
     if (MPIDI_OFI_ENABLE_SCALABLE_ENDPOINTS) {
-        struct fi_cq_attr cq_attr;
+        MPIDI_OFI_CALL(fi_scalable_ep(domain, prov_use, &ep, NULL), ep);
+        MPIDI_OFI_CALL(fi_scalable_ep_bind(ep, &av->fid, 0), bind);
+        MPIDI_OFI_CALL(fi_enable(ep), ep_enable);
 
-        if (*ep == NULL) {
-            /* First call to this function -- set up scalable endpoint */
-            MPIDI_OFI_CALL(fi_scalable_ep(domain, prov_use, ep, NULL), ep);
-            MPIDI_OFI_CALL(fi_scalable_ep_bind(*ep, &av->fid, 0), bind);
+        mpi_errno = create_sep_tx(ep, 0, &tx, cq, rma_cmpl_cntr);
+        MPIR_ERR_CHECK(mpi_errno);
+        mpi_errno = create_sep_rx(ep, 0, &rx, cq, rma_cmpl_cntr);
+        MPIR_ERR_CHECK(mpi_errno);
+    } else {
+        MPIDI_OFI_CALL(fi_endpoint(domain, prov_use, &ep, NULL), ep);
+        MPIDI_OFI_CALL(fi_ep_bind(ep, &av->fid, 0), bind);
+        MPIDI_OFI_CALL(fi_ep_bind(ep, &cq->fid, FI_SEND | FI_RECV | FI_SELECTIVE_COMPLETION), bind);
+        MPIDI_OFI_CALL(fi_ep_bind(ep, &rma_cmpl_cntr->fid, FI_READ | FI_WRITE), bind);
+        MPIDI_OFI_CALL(fi_enable(ep), ep_enable);
+        tx = ep;
+        rx = ep;
+    }
+    MPIDI_OFI_global.ctx[vni].domain = domain;
+    MPIDI_OFI_global.ctx[vni].av = av;
+    MPIDI_OFI_global.ctx[vni].rma_cmpl_cntr = rma_cmpl_cntr;
+    MPIDI_OFI_global.ctx[vni].ep = ep;
+    MPIDI_OFI_global.ctx[vni].cq = cq;
+    MPIDI_OFI_global.ctx[vni].tx = tx;
+    MPIDI_OFI_global.ctx[vni].rx = rx;
+
+#else /* MPIDI_OFI_VNI_USE_SEPCTX */
+    if (vni == 0) {
+        mpi_errno = create_vni_domain(&domain, &av, &rma_cmpl_cntr);
+        MPIR_ERR_CHECK(mpi_errno);
+    } else {
+        domain = MPIDI_OFI_global.ctx[0].domain;
+        av = MPIDI_OFI_global.ctx[0].av;
+        rma_cmpl_cntr = MPIDI_OFI_global.ctx[0].rma_cmpl_cntr;
+    }
+    mpi_errno = create_cq(domain, &cq);
+    MPIR_ERR_CHECK(mpi_errno);
+
+    if (vni == 0) {
+        if (MPIDI_OFI_ENABLE_SCALABLE_ENDPOINTS) {
+            MPIDI_OFI_CALL(fi_scalable_ep(domain, prov_use, &ep, NULL), ep);
+            MPIDI_OFI_CALL(fi_scalable_ep_bind(ep, &av->fid, 0), bind);
+            MPIDI_OFI_CALL(fi_enable(ep), ep_enable);
+        } else {
+            MPIDI_OFI_CALL(fi_endpoint(domain, prov_use, &ep, NULL), ep);
+            MPIDI_OFI_CALL(fi_ep_bind(ep, &av->fid, 0), bind);
+            MPIDI_OFI_CALL(fi_ep_bind(ep, &cq->fid, FI_SEND | FI_RECV | FI_SELECTIVE_COMPLETION),
+                           bind);
+            MPIDI_OFI_CALL(fi_ep_bind(ep, &rma_cmpl_cntr->fid, FI_READ | FI_WRITE), bind);
+            MPIDI_OFI_CALL(fi_enable(ep), ep_enable);
         }
+    } else {
+        ep = MPIDI_OFI_global.ctx[0].ep;
```

Comment:
So for scalable ep path, `ctx[1...N-1].ep;` are all invalid, right? It would be nice to mention that in acomment.

Reply:
Agreed. Will do.